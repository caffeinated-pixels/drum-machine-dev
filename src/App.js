import React from 'react'
import './index.scss'

// SoundBank arrays
const synthwave = [
  {
    trigger: 'Q',
    keycode: 81,
    pad: 'padQ',
    name: 'Kick-1',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/synthwave/Kick-1.mp3'
  },
  {
    trigger: 'W',
    keycode: 87,
    pad: 'padW',
    name: 'Kick-2',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/synthwave/Kick-2.mp3'
  },
  {
    trigger: 'E',
    keycode: 69,
    pad: 'padE',
    name: 'Snare',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/synthwave/Snare.mp3'
  },
  {
    trigger: 'A',
    keycode: 65,
    pad: 'padA',
    name: 'Tom-1',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/synthwave/Tom-1.mp3'
  },
  {
    trigger: 'S',
    keycode: 83,
    pad: 'padS',
    name: 'Tom-2',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/synthwave/Tom-2.mp3'
  },
  {
    trigger: 'D',
    keycode: 68,
    pad: 'padD',
    name: 'Hihat-closed',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/synthwave/Hihat-closed.mp3'
  },
  {
    trigger: 'Z',
    keycode: 90,
    pad: 'padZ',
    name: 'Cymbal-crash',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/synthwave/Cymbol-Crash.mp3'
  },
  {
    trigger: 'X',
    keycode: 88,
    pad: 'padX',
    name: 'Clap',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/synthwave/Clap.mp3'
  },
  {
    trigger: 'C',
    keycode: 67,
    pad: 'padC',
    name: 'Clap-snare',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/synthwave/Clap%2BSnare.mp3'
  }
]

const acoustic = [
  {
    trigger: 'Q',
    keycode: 81,
    pad: 'padQ',
    name: 'Kick-1',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/acoustic/Kick-1.mp3'
  },
  {
    trigger: 'W',
    keycode: 87,
    pad: 'padW',
    name: 'Kick-2',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/acoustic/Kick-2.mp3'
  },
  {
    trigger: 'E',
    keycode: 69,
    pad: 'padE',
    name: 'Snare-1',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/acoustic/Snare-1.mp3'
  },
  {
    trigger: 'A',
    keycode: 65,
    pad: 'padA',
    name: 'Snare-2',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/acoustic/Snare-2.mp3'
  },
  {
    trigger: 'S',
    keycode: 83,
    pad: 'padS',
    name: 'Tom-1',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/acoustic/Tom-1.mp3'
  },
  {
    trigger: 'D',
    keycode: 68,
    pad: 'padD',
    name: 'Tom-2',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/acoustic/Tom-2.mp3'
  },
  {
    trigger: 'Z',
    keycode: 90,
    pad: 'padZ',
    name: 'Hihat-closed',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/acoustic/Hihat-closed.mp3'
  },
  {
    trigger: 'X',
    keycode: 88,
    pad: 'padX',
    name: 'Hihat-open',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/acoustic/Hihat-open.mp3'
  },
  {
    trigger: 'C',
    keycode: 67,
    pad: 'padC',
    name: 'Cymbal-crash',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/acoustic/Cymbal-crash.mp3'
  }
]

const chaosEngine = [
  {
    trigger: 'Q',
    keycode: 81,
    pad: 'padQ',
    name: 'Node-activated',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/chaosengine/Node%20activated.mp3'
  },
  {
    trigger: 'W',
    keycode: 87,
    pad: 'padW',
    name: 'Food',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/chaosengine/Food.mp3'
  },
  {
    trigger: 'E',
    keycode: 69,
    pad: 'padE',
    name: 'Power-up',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/chaosengine/Power%20up.mp3'
  },
  {
    trigger: 'A',
    keycode: 65,
    pad: 'padA',
    name: 'Player-saved',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/chaosengine/Player%20saved.mp3'
  },
  {
    trigger: 'S',
    keycode: 83,
    pad: 'padS',
    name: 'Level-complete',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/chaosengine/Level%20complete.mp3'
  },
  {
    trigger: 'D',
    keycode: 68,
    pad: 'padD',
    name: 'Extra-life',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/chaosengine/Extra%20life.mp3'
  },
  {
    trigger: 'Z',
    keycode: 90,
    pad: 'padZ',
    name: 'Exit-open',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/chaosengine/Exit%20open.mp3'
  },
  {
    trigger: 'X',
    keycode: 88,
    pad: 'padX',
    name: 'Party-power',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/chaosengine/Party%20Power.mp3'
  },
  {
    trigger: 'C',
    keycode: 67,
    pad: 'padC',
    name: 'Shield',
    src:
      'https://raw.githubusercontent.com/caffeinated-pixels/drum-samples/main/chaosengine/Shield.mp3'
  }
]

function TopBar() {
  return (
    <div className="top-bar">
      <p>
        Mercenary <span className="thin">Instruments</span>{' '}
        <span className="model">LV-426</span>
      </p>
      <img
        className="logo"
        src="https://thecakeisaliegaming.files.wordpress.com/2020/11/mercenary-32x32-1.png"
        alt="Mercenary Instruments logo"
      />
    </div>
  )
}

function BankButtons(props) {
  return (
    <div className="btn-container">
      <button
        id="bank1"
        className="bank-btn"
        onClick={props.handleBankBtnClick}
      >
        Bank 1
      </button>
      <button
        id="bank2"
        className="bank-btn"
        onClick={props.handleBankBtnClick}
      >
        Bank 2
      </button>
      <button
        id="bank3"
        className="bank-btn"
        onClick={props.handleBankBtnClick}
      >
        Bank 3
      </button>
    </div>
  )
}

function OtherButtons(props) {
  return (
    <div className="otherbtn-container">
      <div className="power-container">
        <button
          id="power"
          className={`${props.power ? 'powbtn-on' : 'powbtn-off'}`}
          onClick={props.handlePowerBtn}
        >
          On | Off
        </button>
      </div>

      <div className="volume-container">
        <button id="vol-up" className="bank-btn" onClick={props.handleVolBtn}>
          Vol+
        </button>
        <button id="vol-down" className="bank-btn" onClick={props.handleVolBtn}>
          Vol-
        </button>
      </div>
    </div>
  )
}

function Display(props) {
  if (props.power) {
    return (
      <div id="display" className="no-select">
        <p>Bank: {props.bankName}</p>
        <p>Sample: {props.display}</p>
        <p>Volume: {props.volume}</p>
      </div>
    )
  } else {
    return <div id="display"></div>
  }
}

function LeftPanel(props) {
  return (
    <div className="left-panel">
      <Display
        bankName={props.bankName}
        display={props.display}
        volume={props.volume}
        power={props.power}
      />
      <BankButtons handleBankBtnClick={props.handleBankBtnClick} />
      <OtherButtons
        handlePowerBtn={props.handlePowerBtn}
        power={props.power}
        handleVolBtn={props.handleVolBtn}
      />
    </div>
  )
}

function RightPanel(props) {
  const drumpads = props.soundBank.map((entry, i) => {
    return (
      <div
        key={i}
        id={entry.name}
        className={`drum-pad ${entry.pad} `}
        onClick={props.handleClick}
      >
        <audio
          id={entry.trigger}
          className="clip"
          src={entry.src}
          preload="auto"
        >
          {' '}
        </audio>
        {entry.trigger}
      </div>
    )
  })

  return <div className="right-panel no-select">{drumpads}</div>
}

// moved state & playSample up so both LP & RP have access via props
class MainBox extends React.Component {
  constructor() {
    super()
    this.state = {
      currentSample: '',
      soundBank: synthwave,
      bankName: 'Synthwave',
      power: true,
      volume: 1
    }
    this.playSample = this.playSample.bind(this)
    this.handleClick = this.handleClick.bind(this)
    this.handleKeyPress = this.handleKeyPress.bind(this)
    this.handleBankBtnClick = this.handleBankBtnClick.bind(this)
    this.handlePowerBtn = this.handlePowerBtn.bind(this)
    this.handleVolBtn = this.handleVolBtn.bind(this)
  }

  componentDidMount() {
    // add event listener for drumpad keypresses
    document.addEventListener('keydown', this.handleKeyPress)
    // add event listener for removing css animation class
    document.addEventListener('animationend', event =>
      document.getElementById(event.target.id).classList.remove('pad-anim')
    )
  }

  playSample(elementId) {
    const element = document.getElementById(elementId)

    // check if power is on first!
    if (this.state.power) {
      // add class to trigger css animation
      element.classList.add('pad-anim')
      element.firstChild.volume = this.state.volume

      // need to use play() method on nested <audio> element
      document.getElementById(elementId).firstChild.play()
      this.setState(prevState => ({ ...prevState, currentSample: elementId }))
    }
  }

  handleClick(event) {
    this.playSample(event.target.id)
  }

  handleKeyPress(event) {
    // use keycode to to get sample name
    const elementId = this.state.soundBank.filter(
      entry => entry.keycode === event.keyCode
    )

    if (elementId[0]) this.playSample(elementId[0].name)
  }

  handleBankBtnClick(event) {
    let { soundBank, bankName } = this.state

    if (event.target.id === 'bank1') {
      soundBank = synthwave
      bankName = 'Synthwave'
    } else if (event.target.id === 'bank2') {
      soundBank = acoustic
      bankName = 'Acoustic'
    } else if (event.target.id === 'bank3') {
      soundBank = chaosEngine
      bankName = 'Chaos Engine'
    }

    this.setState(prevState => ({
      ...prevState,
      soundBank: soundBank,
      bankName: bankName,
      currentSample: ''
    }))
  }

  handlePowerBtn() {
    // flip power state boolean
    this.setState(prevState => ({
      ...prevState,
      power: !prevState.power,
      currentSample: ''
    }))
  }

  handleVolBtn(event) {
    // find out which vol btn was pressed
    const whichBtn = event.target.id
    this.setState(prevState => {
      if (whichBtn === 'vol-up' && prevState.volume < 1) {
        return {
          ...prevState,
          volume: Number((prevState.volume + 0.1).toFixed(1))
        }
      } else if (whichBtn === 'vol-down' && prevState.volume >= 0.1) {
        return {
          ...prevState,
          volume: Number((prevState.volume - 0.1).toFixed(1))
        }
      }
      /* because of floating point precision errors we need to round the return val to 1 decimal place; tofixed() returns a string, so we convert back to a num using Number() */
    })
  }

  render() {
    return (
      <div className="main-box">
        <LeftPanel
          display={this.state.currentSample}
          handleBankBtnClick={this.handleBankBtnClick}
          bankName={this.state.bankName}
          handlePowerBtn={this.handlePowerBtn}
          power={this.state.power}
          volume={this.state.volume}
          handleVolBtn={this.handleVolBtn}
        />
        <RightPanel
          handleClick={this.handleClick}
          soundBank={this.state.soundBank}
          activePad={this.state.activePad}
          power={this.state.power}
        />
      </div>
    )
  }
}

function BottomBar() {
  return (
    <div className="bottom-bar text-center">
      <p>
        Coded & abused by&nbsp;
        <a
          href="https://steviegill-webportfolio.netlify.app/"
          target="_blank"
          rel="noreferrer"
        >
          Stevie Gill
        </a>
      </p>
    </div>
  )
}

function App() {
  return (
    <div id="drum-machine">
      <TopBar />
      <MainBox />
      <BottomBar />
    </div>
  )
}

export default App
