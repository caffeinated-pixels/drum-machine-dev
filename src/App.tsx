import { TopBar, MainBox, BottomBar } from './components'

export const App = () => {
  return (
    <div id="drum-machine">
      <TopBar />
      <MainBox />
      <BottomBar />
    </div>
  )
}
