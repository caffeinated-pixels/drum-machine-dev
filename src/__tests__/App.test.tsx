import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import renderer from 'react-test-renderer'
import userEvent from '@testing-library/user-event'
import { App } from '../App'
import {
  BANK_NAMES,
  BUTTON_NAMES,
  PAD_LABELS,
  SCREEN_LABELS,
} from '../constants/names'

describe('Drum machine app', () => {
  it('Matches DOM tree snapshot', () => {
    const domTree = renderer.create(<App />).toJSON()
    expect(domTree).toMatchSnapshot()
  })

  it('should render the display', () => {
    render(<App />)
  })
})

describe('display', () => {
  it('should update when one of the bank buttons is clicked', async () => {
    render(<App />)
    const display = screen.getByTestId('display')
    const bank1Btn = screen.getByRole('button', { name: BUTTON_NAMES.BANK1 })
    const bank2Btn = screen.getByRole('button', { name: BUTTON_NAMES.BANK2 })
    const bank3Btn = screen.getByRole('button', { name: BUTTON_NAMES.BANK3 })

    expect(display).toHaveTextContent(BANK_NAMES.BANK1)
    await userEvent.click(bank2Btn)
    expect(display).toHaveTextContent(BANK_NAMES.BANK2)
    await userEvent.click(bank3Btn)
    expect(display).toHaveTextContent(BANK_NAMES.BANK3)
    await userEvent.click(bank1Btn)
    expect(display).toHaveTextContent(BANK_NAMES.BANK1)
  })

  it('should update when the volume buttons are clicked', async () => {
    render(<App />)
    const volumeDisplay = screen.getByText(/volume/i)
    const increaseVol = screen.getByRole('button', {
      name: BUTTON_NAMES.VOL_UP,
    })
    const decreaseVol = screen.getByRole('button', {
      name: BUTTON_NAMES.VOL_DOWN,
    })

    expect(volumeDisplay).toHaveTextContent(`${SCREEN_LABELS.VOLUME}: 1`)
    await userEvent.click(decreaseVol)
    expect(volumeDisplay).toHaveTextContent(`${SCREEN_LABELS.VOLUME}: 0.9`)
    await userEvent.click(decreaseVol)
    expect(volumeDisplay).toHaveTextContent(`${SCREEN_LABELS.VOLUME}: 0.8`)
    await userEvent.click(decreaseVol)
    expect(volumeDisplay).toHaveTextContent(`${SCREEN_LABELS.VOLUME}: 0.7`)
    await userEvent.click(increaseVol)
    expect(volumeDisplay).toHaveTextContent(`${SCREEN_LABELS.VOLUME}: 0.8`)
    await userEvent.click(increaseVol)
    expect(volumeDisplay).toHaveTextContent(`${SCREEN_LABELS.VOLUME}: 0.9`)
  })

  it('voulme should not exceed 1', async () => {
    render(<App />)
    const volumeDisplay = screen.getByText(/volume/i)
    const increaseVol = screen.getByRole('button', {
      name: BUTTON_NAMES.VOL_UP,
    })
    const decreaseVol = screen.getByRole('button', {
      name: BUTTON_NAMES.VOL_DOWN,
    })

    expect(volumeDisplay).toHaveTextContent(`${SCREEN_LABELS.VOLUME}: 1`)
    await userEvent.click(increaseVol)
    expect(volumeDisplay).toHaveTextContent(`${SCREEN_LABELS.VOLUME}: 1`)
    await userEvent.click(decreaseVol)
    expect(volumeDisplay).toHaveTextContent(`${SCREEN_LABELS.VOLUME}: 0.9`)
  })

  it('voulme should not go below 0', async () => {
    render(<App />)
    const volumeDisplay = screen.getByText(/volume/i)
    const decreaseVol = screen.getByRole('button', {
      name: BUTTON_NAMES.VOL_DOWN,
    })
    const increaseVol = screen.getByRole('button', {
      name: BUTTON_NAMES.VOL_UP,
    })

    await userEvent.tripleClick(decreaseVol)
    await userEvent.tripleClick(decreaseVol)
    await userEvent.tripleClick(decreaseVol)
    await userEvent.tripleClick(decreaseVol)
    expect(volumeDisplay).toHaveTextContent(`${SCREEN_LABELS.VOLUME}: 0`)
    await userEvent.click(increaseVol)
    expect(volumeDisplay).toHaveTextContent(`${SCREEN_LABELS.VOLUME}: 0.1`)
  })

  it('should be blank when power turned off', async () => {
    render(<App />)

    const powerButton = screen.getByRole('button', { name: BUTTON_NAMES.POWER })
    expect(screen.getByText(/Bank: Synthwave/)).toBeInTheDocument()
    await userEvent.click(powerButton)
    expect(screen.queryByText(/Bank: Synthwave/)).not.toBeInTheDocument()
  })

  // TODO: mock Howler to test audio & prevent console errors
  it('should display previous values when turned back on', async () => {
    render(<App />)

    const bank3Btn = screen.getByRole('button', { name: BUTTON_NAMES.BANK3 })
    await userEvent.click(bank3Btn)
    const padW = screen.getByRole('button', { name: PAD_LABELS.W })
    await userEvent.click(padW)
    const decreaseVol = screen.getByRole('button', {
      name: BUTTON_NAMES.VOL_DOWN,
    })
    await userEvent.tripleClick(decreaseVol)
    const powerButton = screen.getByRole('button', { name: BUTTON_NAMES.POWER })
    await userEvent.click(powerButton)
    await userEvent.click(powerButton)

    expect(screen.getByText(`Bank: ${BANK_NAMES.BANK3}`)).toBeInTheDocument()
  })
})
