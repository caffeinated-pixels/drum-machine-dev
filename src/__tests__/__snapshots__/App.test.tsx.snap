// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator app Matches DOM tree snapshot 1`] = `
<div
  id="drum-machine"
>
  <div
    className="top-bar"
  >
    <p>
      Mercenary 
      <span
        className="thin"
      >
        Instruments
      </span>
       
      <span
        className="model"
      >
        LV-426
      </span>
    </p>
    <img
      alt="Mercenary Instruments logo"
      className="logo"
      src="./favicon.png"
    />
  </div>
  <div
    className="main-box"
  >
    <div
      className="left-panel"
    >
      <div
        className="display-container"
      >
        <div
          className="no-select"
          id="display"
        >
          <p>
            Bank: 
            Synthwave
          </p>
          <p>
            Sample: 
          </p>
          <p>
            Volume: 
            1
          </p>
        </div>
      </div>
      <div
        className="btn-container"
      >
        <button
          className="bank-btn"
          id="bank1"
          onClick={[Function]}
        >
          Bank 1
        </button>
        <button
          className="bank-btn"
          id="bank2"
          onClick={[Function]}
        >
          Bank 2
        </button>
        <button
          className="bank-btn"
          id="bank3"
          onClick={[Function]}
        >
          Bank 3
        </button>
      </div>
      <div
        className="otherbtn-container"
      >
        <div
          className="power-container"
        >
          <button
            className="powbtn-on"
            id="power"
            onClick={[Function]}
          >
            On | Off
          </button>
        </div>
        <div
          className="volume-container"
        >
          <button
            className="bank-btn"
            id="vol-up"
            onClick={[Function]}
          >
            Vol+
          </button>
          <button
            className="bank-btn"
            id="vol-down"
            onClick={[Function]}
          >
            Vol-
          </button>
        </div>
      </div>
    </div>
    <div
      className="right-panel no-select"
    >
      <button
        className="drum-pad padQ "
        id="Kick-1"
        onClick={[Function]}
      >
        <audio
          className="clip"
          id="Q"
          preload="auto"
          src="kick1.mp3"
        >
           
        </audio>
        Q
      </button>
      <button
        className="drum-pad padW "
        id="Kick-2"
        onClick={[Function]}
      >
        <audio
          className="clip"
          id="W"
          preload="auto"
          src="kick2.mp3"
        >
           
        </audio>
        W
      </button>
      <button
        className="drum-pad padE "
        id="Snare"
        onClick={[Function]}
      >
        <audio
          className="clip"
          id="E"
          preload="auto"
          src="snare.mp3"
        >
           
        </audio>
        E
      </button>
      <button
        className="drum-pad padA "
        id="Tom-1"
        onClick={[Function]}
      >
        <audio
          className="clip"
          id="A"
          preload="auto"
          src="tom1.mp3"
        >
           
        </audio>
        A
      </button>
      <button
        className="drum-pad padS "
        id="Tom-2"
        onClick={[Function]}
      >
        <audio
          className="clip"
          id="S"
          preload="auto"
          src="tom2.mp3"
        >
           
        </audio>
        S
      </button>
      <button
        className="drum-pad padD "
        id="Hihat-closed"
        onClick={[Function]}
      >
        <audio
          className="clip"
          id="D"
          preload="auto"
          src="hihatClosed.mp3"
        >
           
        </audio>
        D
      </button>
      <button
        className="drum-pad padZ "
        id="Cymbal-crash"
        onClick={[Function]}
      >
        <audio
          className="clip"
          id="Z"
          preload="auto"
          src="cymbalCrash.mp3"
        >
           
        </audio>
        Z
      </button>
      <button
        className="drum-pad padX "
        id="Clap"
        onClick={[Function]}
      >
        <audio
          className="clip"
          id="X"
          preload="auto"
          src="clap.mp3"
        >
           
        </audio>
        X
      </button>
      <button
        className="drum-pad padC "
        id="Clap-snare"
        onClick={[Function]}
      >
        <audio
          className="clip"
          id="C"
          preload="auto"
          src="clapSnare.mp3"
        >
           
        </audio>
        C
      </button>
    </div>
  </div>
  <div
    className="bottom-bar text-center"
  >
    <p>
      Coded byÂ 
      <a
        href="https://steviegill-webportfolio.netlify.app/"
        title="Stevie's web app portofolio page"
      >
        Stevie Gill
      </a>
      ; 
      <a
        href="https://github.com/caffeinated-pixels/drum-machine-dev"
        title="Drum machine Github repo"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-square-github githubIcon"
          data-icon="square-github"
          data-prefix="fab"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zM277.3 415.7c-8.4 1.5-11.5-3.7-11.5-8 0-5.4.2-33 .2-55.3 0-15.6-5.2-25.5-11.3-30.7 37-4.1 76-9.2 76-73.1 0-18.2-6.5-27.3-17.1-39 1.7-4.3 7.4-22-1.7-45-13.9-4.3-45.7 17.9-45.7 17.9-13.2-3.7-27.5-5.6-41.6-5.6-14.1 0-28.4 1.9-41.6 5.6 0 0-31.8-22.2-45.7-17.9-9.1 22.9-3.5 40.6-1.7 45-10.6 11.7-15.6 20.8-15.6 39 0 63.6 37.3 69 74.3 73.1-4.8 4.3-9.1 11.7-10.6 22.3-9.5 4.3-33.8 11.7-48.3-13.9-9.1-15.8-25.5-17.1-25.5-17.1-16.2-.2-1.1 10.2-1.1 10.2 10.8 5 18.4 24.2 18.4 24.2 9.7 29.7 56.1 19.7 56.1 19.7 0 13.9.2 36.5.2 40.6 0 4.3-3 9.5-11.5 8-66-22.1-112.2-84.9-112.2-158.3 0-91.8 70.2-161.5 162-161.5S388 165.6 388 257.4c.1 73.4-44.7 136.3-110.7 158.3zm-98.1-61.1c-1.9.4-3.7-.4-3.9-1.7-.2-1.5 1.1-2.8 3-3.2 1.9-.2 3.7.6 3.9 1.9.3 1.3-1 2.6-3 3zm-9.5-.9c0 1.3-1.5 2.4-3.5 2.4-2.2.2-3.7-.9-3.7-2.4 0-1.3 1.5-2.4 3.5-2.4 1.9-.2 3.7.9 3.7 2.4zm-13.7-1.1c-.4 1.3-2.4 1.9-4.1 1.3-1.9-.4-3.2-1.9-2.8-3.2.4-1.3 2.4-1.9 4.1-1.5 2 .6 3.3 2.1 2.8 3.4zm-12.3-5.4c-.9 1.1-2.8.9-4.3-.6-1.5-1.3-1.9-3.2-.9-4.1.9-1.1 2.8-.9 4.3.6 1.3 1.3 1.8 3.3.9 4.1zm-9.1-9.1c-.9.6-2.6 0-3.7-1.5s-1.1-3.2 0-3.9c1.1-.9 2.8-.2 3.7 1.3 1.1 1.5 1.1 3.3 0 4.1zm-6.5-9.7c-.9.9-2.4.4-3.5-.6-1.1-1.3-1.3-2.8-.4-3.5.9-.9 2.4-.4 3.5.6 1.1 1.3 1.3 2.8.4 3.5zm-6.7-7.4c-.4.9-1.7 1.1-2.8.4-1.3-.6-1.9-1.7-1.5-2.6.4-.6 1.5-.9 2.8-.4 1.3.7 1.9 1.8 1.5 2.6z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
         repo
      </a>
    </p>
  </div>
</div>
`;
